core/src/mender-client.c

```diff
@@ -1120,109 +1115,288 @@ mender_client_update_work_function(void) {
+    mender_update_state_t         update_state            = MENDER_UPDATE_STATE_DOWNLOAD;
+    const char                   *deployment_id;
+
+    /* reset the currently used update module */
+    mender_client_update_module = NULL;
+
+    {
+        char *artifact_type;
+        if (MENDER_OK == (ret = mender_storage_get_update_state(&update_state, &artifact_type))) {
+            mender_client_update_module = mender_client_get_update_module(artifact_type);
+            free(artifact_type);
         }
-        goto END;
     }
-    cJSON_AddStringToObject(mender_client_deployment_data, "provides", new_provides);
-#endif
-#endif
 
-    /* Set boot partition */
-    mender_log_info("Download done, installing artifact");
-    mender_client_publish_deployment_status(deployment->id, MENDER_DEPLOYMENT_STATUS_INSTALLING);
-    if (true == mender_client_deployment_needs_set_pending_image) {
-        if (MENDER_OK != (ret = mender_flash_set_pending_image(mender_client_flash_handle))) {
-            mender_log_error("Unable to set boot partition");
-            mender_client_publish_deployment_status(deployment->id, MENDER_DEPLOYMENT_STATUS_FAILURE);
+    /* Skip the block below if we just resume from a saved state. */
+    /* TODO: move this into the state code */
+    if (update_state == MENDER_UPDATE_STATE_DOWNLOAD) {
+        mender_log_info("Checking for deployment...");
+        if (MENDER_OK != (ret = mender_api_check_for_deployment(deployment))) {
+            mender_log_error("Unable to check for deployment");
             goto END;
         }
-    }
 
-    /* Check if the system must restart following downloading the deployment */
-    if (true == mender_client_deployment_needs_restart) {
-        /* Save deployment data to publish deployment status after rebooting */
-        if (NULL == (storage_deployment_data = cJSON_PrintUnformatted(mender_client_deployment_data))) {
-            mender_log_error("Unable to save deployment data");
-            mender_client_publish_deployment_status(deployment->id, MENDER_DEPLOYMENT_STATUS_FAILURE);
-            ret = MENDER_FAIL;
+        /* Check if deployment is available */
+        if ((NULL == deployment->id) || (NULL == deployment->artifact_name) || (NULL == deployment->uri) || (NULL == deployment->device_types_compatible)) {
+            mender_log_info("No deployment available");
+            ret = MENDER_DONE;
             goto END;
         }
-        if (MENDER_OK != (ret = mender_storage_set_deployment_data(storage_deployment_data))) {
-            mender_log_error("Unable to save deployment data");
-            mender_client_publish_deployment_status(deployment->id, MENDER_DEPLOYMENT_STATUS_FAILURE);
-            goto END;
+
+        /* Reset flags */
+        mender_client_deployment_needs_set_pending_image = false;
+
+        /* Create deployment data */
+        if (NULL != mender_client_deployment_data) {
+            cJSON_Delete(mender_client_deployment_data);
```

Comment:
It's just a "better safe than sorry" way to avoid a memory leak in case there is `deployment_data` from somewhere and we get to this point. I can add an `assert()` if it makes you feel better (and you don't have enough crashes :wink: )

Reply:
A warning would be nice. Leave the assert out for now